src/Interpreters/ConcurrentHashJoin.cpp

```diff
@@ -0,0 +1,323 @@
+#include <memory>
+#include <mutex>
+#include <Columns/FilterDescription.h>
+#include <Columns/IColumn.h>
+#include <Core/ColumnsWithTypeAndName.h>
+#include <Core/NamesAndTypes.h>
+#include <IO/Operators.h>
+#include <IO/WriteBufferFromString.h>
+#include <Interpreters/ActionsDAG.h>
+#include <Interpreters/ActionsVisitor.h>
+#include <Interpreters/ConcurrentHashJoin.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/ExpressionActions.h>
+#include <Interpreters/PreparedSets.h>
+#include <Interpreters/SubqueryForSet.h>
+#include <Interpreters/TableJoin.h>
+#include <Parsers/DumpASTNode.h>
+#include <Parsers/ExpressionListParsers.h>
+#include <Parsers/IAST_fwd.h>
+#include <Parsers/parseQuery.h>
+#include <Common/Exception.h>
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int SET_SIZE_LIMIT_EXCEEDED;
+    extern const int BAD_ARGUMENTS;
+}
+namespace JoinStuff
+{
+ConcurrentHashJoin::ConcurrentHashJoin(ContextPtr context_, std::shared_ptr<TableJoin> table_join_, size_t slots_, const Block & right_sample_block, bool any_take_last_row_)
+    : context(context_)
+    , table_join(table_join_)
+    , slots(slots_)
+{
+    if (!slots_)
+    {
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Invalid argument slot : {}", slots_);
+    }
+
+    for (size_t i = 0; i < slots; ++i)
+    {
+        auto inner_hash_join = std::make_shared<InnerHashJoin>();
+        inner_hash_join->data = std::make_unique<HashJoin>(table_join_, right_sample_block, any_take_last_row_);
+        hash_joins.emplace_back(std::move(inner_hash_join));
+    }
+    dispatch_datas.emplace_back(std::make_shared<BlockDispatchControlData>());
+    dispatch_datas.emplace_back(std::make_shared<BlockDispatchControlData>());
+}
+
+bool ConcurrentHashJoin::addJoinedBlock(const Block & block, bool check_limits)
+{
+    auto & dispatch_data = getBlockDispatchControlData(block, RIGHT);
+    std::vector<Block> dispatched_blocks;
+    Block cloned_block = block;
+    dispatchBlock(dispatch_data, cloned_block, dispatched_blocks);
+    for (size_t i = 0; i < dispatched_blocks.size(); ++i)
+    {
+        auto & hash_join = hash_joins[i];
+        auto & dispatched_block = dispatched_blocks[i];
+        auto rows = dispatched_block.rows();
+        check_total_rows += rows;
+        check_total_bytes += dispatched_block.bytes();
+        std::unique_lock lock(hash_join->mutex);
+        
+        // Don't take the real insertion here, because inserting a block into HashTable is a time-consuming operation,
+        // it may cause serious lock contention and make the whole process slow.
+        hash_join->pending_right_blocks.emplace_back(std::move(dispatched_block));
```

Comment:
Did you test implementation without this queue? it seems a bit over-complicated. Here we have separate mutex for each `hash_join`, so calls `ConcurrentHashJoin::addJoinedBlock` would block each other only when accessing same `hash_join` instance. Maybe make one iteration with `try_lock` to insert into available slots and then another loop with waiting.

Reply:
Maybe it takes more time, but We could have a test on this